docs/protocol/papers/blockchain-programs.md

```diff
@@ -209,20 +210,20 @@ When the program is instantiated with a `publicKey` value — to be used in an o
 When the output is spent and the control program is run:
 
 1. The virtual machine first takes arguments specified in the input witness and pushes them to the stack. In this case, that argument is a signature.
-2. The program then executes, first pushing the public key and then the transaction hash to the stack. 
-3. The public key and transaction hash are then swapped to put them in the correct order for the following `CHECKSIG` instruction. 
-4. `CHECKSIG` pops all three items off the stack to check the signature, pushing `true` or `false` to the stack. 
+2. The program then executes, first pushing the public key and then the transaction hash to the stack.
+3. The public key and transaction hash are then swapped to put them in the correct order for the following `CHECKSIG` instruction.
+4. `CHECKSIG` pops all three items off the stack to check the signature, pushing `true` or `false` to the stack.
 5. `VERIFY` then pops the top value from the stack and causes the program to fail if the value is `false`. (In an actual control program, the `VERIFY` instruction of the last condition in a path is omitted, since it is performed by the VM itself.)
 
 Many control, issuance, and consensus programs use a multisignature check.
 
-	program MultiKeyProgram(n, m, publicKeys[n]) {
-		path spend(signatures[m]) {
+	program MultiKeyProgram(n: Number, m: Number, publicKeys: PublicKey[n]) {
```

Comment:
You're right that this is a form of dependent typing. It's a hint that's unfortunately necessary given the way lists are currently represented (as multiple stack items rather than a single list). We can't just do an implicit carried `n` because sometimes the length of an array is specified in the program arguments but the array itself is provided in the path arguments (which is true for the list of signatures). But since this is something that we're planning to fix in the VM, maybe we should just do the right thing. Since the previous solution (`publicKeys[m]`) is approximately equally weird, could we land this and fix this problem in a separate PR if we come up with something?